src/ipld/frozen_cids.rs

```diff
@@ -0,0 +1,248 @@
+// (which can be stored in 32 bytes vs 96 bytes for a `Cid` type) is +99.99% of
+// all CIDs. `FrozenCids` defaults to the `Box<[u8; BLAKE2B256_SIZE]>` variant of
+// `CidBox`, only using the more expensive `Box<Cid>` variant when necessary.
+// The Box type has been chosen to make `FrozenCids` explicitly immutable.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FrozenCids(CidBox);
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        Self(CidBox::V1DagCborBlake2bCids(Box::new([])))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum CidBox {
+    V1DagCborBlake2bCids(Box<[[u8; BLAKE2B256_SIZE]]>),
+    AllCids(Box<[Cid]>),
+}
+
+pub struct FrozenCidsIterator<'a> {
+    buffer: &'a FrozenCids,
+    current_ix: usize,
+}
+
+impl<'a> IntoIterator for &'a FrozenCids {
+    type Item = Cid;
+    type IntoIter = FrozenCidsIterator<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        FrozenCidsIterator {
+            buffer: self,
+            current_ix: 0,
+        }
+    }
+}
+
+impl Iterator for FrozenCidsIterator<'_> {
+    type Item = Cid;
+    fn next(&mut self) -> Option<Self::Item> {
+        match &self.buffer.0 {
+            CidBox::V1DagCborBlake2bCids(cids) => {
+                if self.current_ix >= cids.len() {
+                    None
+                } else {
+                    let cid = Cid::new_v1(
+                        DAG_CBOR,
+                        multihash::Multihash::wrap(Blake2b256.into(), &cids[self.current_ix])
+                            .expect("failed to convert Blake2b digest to V1 DAG-CBOR Blake2b CID"),
+                    );
+                    self.current_ix += 1;
+                    Some(cid)
+                }
+            }
+            CidBox::AllCids(cids) => {
+                if self.current_ix >= cids.len() {
+                    None
+                } else {
+                    let cid = cids[self.current_ix];
+                    self.current_ix += 1;
+                    Some(cid)
+                }
+            }
+        }
+    }
+}
+
+impl Serialize for FrozenCids {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        Vec::<Cid>::from(self).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for FrozenCids {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self::from(<Vec<Cid>>::deserialize(deserializer)?))
+    }
+}
+
+impl FromIterator<Cid> for FrozenCids {
+    fn from_iter<T: IntoIterator<Item = Cid>>(iter: T) -> Self {
+        let mut vec = Vec::new();
+        for i in iter {
+            vec.push(i);
+        }
+        FrozenCids::from(vec)
+    }
+}
+
+// Converts `Vec<Cid>` to `FrozenCids(CidBox::V1DagCborBlake2bCids)` if possible; otherwise, converts to `FrozenCids(CidBox::AllCids)`.
+impl From<Vec<Cid>> for FrozenCids {
+    fn from(cids: Vec<Cid>) -> Self {
```

Comment:
This loses ordering, which I think is unexpected

Reply:
Corrected.